### PR TITLE
Refactor flash and erase error handling

### PR DESCRIPTION
**Pull request by juhhov**
## Status
**READY**

## Description
Refactor FlasherMbed error handling to more pythonic way. Command line error handling is kept backwards compatible. Eventually all flashers should be changed to similar approach.
Add retry for DAPLink transient and user errors.

[IOTSYSTOOL-687](https://jira.arm.com/browse/IOTSYSTOOL-687)
[IOTSYSTOOL-698](https://jira.arm.com/browse/IOTSYSTOOL-698)

## Todos
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: mbed_flasher/common.py [hunk 1]
```diff
@@ -110,11 +110,22 @@ def retry(logger, func, func_args, retries=DEFAULT_RETRY_AMOUNT, conditions=None
...
 
     return return_code
+
+
+class FlashError(Exception):
+    def __init__(self, message, return_code):
+        super(FlashError, self).__init__(message)
+        self.return_code = return_code
+
+
```
Comment (jonikula) [inline mbed_flasher/common.py]: Extra newline here at the end?

File: mbed_flasher/flashers/FlasherMbed.py [hunk 2]
```diff
@@ -153,14 +157,16 @@ def flash(self, source, target, method, no_reset):
...
+                             return_code=EXIT_CODE_EGDB_NOT_SUPPORTED)
 
         return retry(
             logger=self.logger,
             func=self.try_drag_and_drop_flash,
             func_args=(source, target, no_reset),
-            conditions=[EXIT_CODE_OS_ERROR])
+            conditions=[EXIT_CODE_OS_ERROR,
+                        EXIT_CODE_DAPLINK_TRANSIENT_ERROR,
+                        EXIT_CODE_DAPLINK_SOFTWARE_ERROR])
```
Comment (YifeiZuo01) [inline mbed_flasher/flashers/FlasherMbed.py]: Does function `retry()` parameter meant to be configurable by user, e.g. clitest?  
if so, it seems not exposed here, maybe also need changes in `flash.py`.

Comment (juhhov) [reply mbed_flasher/flashers/FlasherMbed.py]: It is not currently meant to be configurable by user.

Comment (YifeiZuo01) [reply mbed_flasher/flashers/FlasherMbed.py]: okay

**Comment by jupe** [comment]
maybe retry change should be moved to another pr

File: test/hardware/test_flash.py [hunk 3]
```diff
@@ -135,24 +137,23 @@ def test_run_fail_file(self, mock_stdout):
...
-                os.system('rm %s' % os.path.join(mount_point, 'failing.txt'))
-                os.system('rm %s' % os.path.join(os.getcwd(), fail_txt_path))
-            self.assertEqual(ret, EXIT_CODE_FILE_STILL_PRESENT)
-        if mock_stdout:
-            pass
+            flasher.flash(source=fail_txt_path, target=target_to_test,
+                          method='simple', no_reset=False)
+
+        if platform.system() == 'Windows':
+            os.system('del %s' % os.path.join(mount_point, 'failing.txt'))
```
Comment (jupe) [inline test/hardware/test_flash.py]: add dedicated function which handle this os condition

File: doc/exit_codes.md [hunk 4]
```diff
@@ -12,7 +12,6 @@
 |16|EXIT_CODE_FILE_DOES_NOT_EXIST|
 |17|EXIT_CODE_FILE_COULD_NOT_BE_READ|
 |18|EXIT_CODE_TARGET_ID_CONFLICT|
-|19|EXIT_CODE_IO_ERROR|
```
Comment (jupe) [inline doc/exit_codes.md]: is this missing from here:

EXIT_CODE_UNHANDLED_EXCEPTION = 1

**Comment by jupe** [comment]
https://github.com/ARMmbed/mbed-flasher/releases/tag/v0.8.0-rc1

**Pull request merged**